### ISSUE TITLE
1/3 - 1/2 size option ??

### ISSUE THREAD

**Issue by eltomjan** [tchapi/Adafruit-GFX-Font-Customiser#39]
What do you think about XOR compresion ?
Tryied to put simple code to export
``` JavaScript
// Set data width/height to 0 for disabled glyphs
const w = parseInt(t.attr('data-w')) * (1 - parseInt(t.attr('data-dis')))
const h = parseInt(t.attr('data-h')) * (1 - parseInt(t.attr('data-dis')))
const hReal = dataPixels.length / w;
if (hReal != parseInt(hReal))
  debugger;
let xored = [dataPixels.substr(0, w)];
for(var y=0;y<h;y++)
{
  var row = [];
  for(var x=0;x<w;x++)
  {
    var pos = y*w + x;
    row.push((dataPixels[pos] != dataPixels[pos + w]) ? 1 : 0);
  }
  xored.push(row.join(""));
}
dataPixels = xored.join("");
```
I can see XORed results by export/import, but there would be need to replace blocks of zeros by 0 bit, prefix 1 before non-zero and recalculate glyph indexes.

Real font example with graphics in comments.
https://github.com/eltomjan/ETEhomeTools/blob/master/ArduinoSimulator/Font_big.h
PR for original LCD
https://github.com/adafruit/Adafruit_ILI9340/pull/19
Here graphical demo showing these blocks inside real font
https://eltomjan.github.io/JStoolsSPAdemos/FontPacker/webLog.htm

I can create PR, but page is quite complex and not sure if there are any "global" changes already ?
Also do not know where to add it to UI - options could be checkbox(-es), new buttons, auto import of packed/plain, etc. ?

PC simulation
https://github.com/eltomjan/ETEhomeTools/tree/master/ArduinoSimulator

Old C# packer to generate fonts
https://github.com/eltomjan/ETEhomeTools/tree/master/CSharp/GFXpack

Some older generator / list of compression results
https://github.com/eltomjan/ETEhomeTools/blob/02c368e44a83ed0c4cb7ae59315b8c681f98fe11/gfxXorPacker.cpp#L332

**Comment by tchapi** [comment]
Hi @eltomjan and thanks for the issue

I need to get my head around what you propose, which is not easy as I'm not versed into XOR compression. I have limited bandwidth right now.

About the UI: a simple checkbox when exporting would do, if you get motivated to propose a PR?

**Comment by eltomjan** [comment]
Ok, will try PR.
Anyway idea is fairy simple...
If you xor each row with previous, you get rid of most 1 bits - just one
horizontal line for vertical like kind of "I" letter, pipe, etc.
And then encode bit stream in a way so that zero block will be zero bit and
non-zero block will have prefix 1.
That is it, during decode recreate blocks (multiply 0-s by block size or
remove 1st 1-s) just xor again to get original.
Have also sketch simulation ArduinoSimulator project for Visual Studio to
play with or debug before flashing...

BR,
Jan

Dne čt 14. 3. 2024 18:03 uživatel Cyril Chapellier ***@***.***>
napsal:

**Comment by replaysMike** [comment]
Just came across this myself, going to test out this PR

**Comment by replaysMike** [comment]
PR works great, tried the hosted [demo](https://eltomjan.github.io/JStoolsSPAdemos/Adafruit-GFX-Font-Customiser.htm) and now my build is 4k smaller! Awesome thanks @eltomjan now I just need to implement the library decompression portion which should be trivial.

**Comment by eltomjan** [comment]
> ... decompression portion which should be trivial.
I did not work on real HW for quite some time yet, but made few demos like here
https://wokwi.com/projects/394882822504014849
Only tricky parts are to get data from FLASH (hard to simulate) and find possible bugs when LCD gets broken :-)

**Comment by replaysMike** [comment]
> > ... decompression portion which should be trivial.
> I did not work on real HW for quite some time yet, but made few demos like here
> https://wokwi.com/projects/394882822504014849
> Only tricky parts are to get data from FLASH (hard to simulate) and find possible bugs when LCD gets broken :-)
> 

Yup I had no issue getting it working yesterday, integrated your decompression code into a custom/trimmed build of Adafruit GFX library. It was a real life saver, I went from having only 4 bytes free on my Attiny3224 to around 4kb free - room for more much needed code. Maybe someday they will release a 64kb attiny 😀
